XenModel/Actions/HealthCheck/TransferXenCenterMetadata.cs

```
@@ -0,0 +1,96 @@
+using System.IO.Pipes;
+using System.ServiceProcess;
+using System.Text;
+using XenAdmin.Core;
+using XenAdmin.Model;
+using XenAPI;
+
+namespace XenAdmin.Actions
+{
+    public class TransferXenCenterMetadata : PureAsyncAction
+    {
+        private string metadata;
+
+        public TransferXenCenterMetadata(string metadata, bool suppressHistory)
+            : base(null, Messages.ACTION_TRANSFER_HEALTHCHECK_SETTINGS, Messages.ACTION_TRANSFER_HEALTHCHECK_SETTINGS, suppressHistory)
+        {
+            this.metadata = metadata;
+        }
+
+        private const string HEALTHCHECKSERVICENAME = "XenServerHealthCheck";
+        private const char SEPARATOR = '\x202f'; // narrow non-breaking space.
+
+        protected override void Run()
+        {
+            try
+            {
+                ServiceController sc = new ServiceController(HEALTHCHECKSERVICENAME);
+                if (sc.Status != ServiceControllerStatus.Running)
+                    return;
+            }
+            catch
+            {
+                return;
+            }
+
+            NamedPipeClientStream pipeClient = new NamedPipeClientStream(".", HealthCheckSettings.HEALTH_CHECK_PIPE, PipeDirection.Out);
+            int retryCount = 120;
+            do
+            {
+                try
+                {
+                    pipeClient.Connect(0);
+                }
+                catch (System.TimeoutException exp)
+                {
+                    throw exp;
+                }
+                catch
+                {
+                    System.Threading.Thread.Sleep(1000);
+                    pipeClient = null;
+                    pipeClient = new NamedPipeClientStream(".", HealthCheckSettings.HEALTH_CHECK_PIPE, PipeDirection.Out);
+                }
+            } while (!pipeClient.IsConnected && retryCount-- != 0);
+
+            byte[] metadataBytes = Encoding.UTF8.GetBytes(String.Join(SEPARATOR.ToString(), HealthCheckSettings.XENCENTER_METADATA, metadata));

```

An old Health Check service will ignore this new message. It will actually try to parse it as a "server list update" and fail silently, because the service expects a server list update message to be an encrypted string (while the metadata message contains the non-encrypted identifier followed by the encrypted metadata).